sync2/multipeer/setsyncbase.go

```diff
@@ -0,0 +1,182 @@
+package multipeer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/sync2/rangesync"
+)
+
+// SetSyncBase is a synchronization base which holds the original OrderedSet.
+// For each peer, a Syncer is derived from the base which is used to synchronize against
+// that peer only. This way, there's no propagation of any keys for which the actual data
+// has not been yet received and validated.
+type SetSyncBase struct {
+	mtx     sync.Mutex
+	ps      PairwiseSyncer
+	os      OrderedSet
+	handler SyncKeyHandler
+	waiting []<-chan singleflight.Result
+	g       singleflight.Group
+}
+
+var _ SyncBase = &SetSyncBase{}
+
+// NewSetSyncBase creates a new SetSyncBase.
+func NewSetSyncBase(ps PairwiseSyncer, os OrderedSet, handler SyncKeyHandler) *SetSyncBase {
+	return &SetSyncBase{
+		ps:      ps,
+		os:      os,
+		handler: handler,
+	}
+}
+
+// Count implements SyncBase.
+func (ssb *SetSyncBase) Count() (int, error) {
+	// TODO: don't lock on db-bound operations
+	ssb.mtx.Lock()
+	defer ssb.mtx.Unlock()
+	if empty, err := ssb.os.Empty(); err != nil {
+		return 0, fmt.Errorf("check if the set is empty: %w", err)
+	} else if empty {
+		return 0, nil
+	}
+	x, err := ssb.os.Items().First()
+	if err != nil {
+		return 0, fmt.Errorf("get first item: %w", err)
+	}
+	info, err := ssb.os.GetRangeInfo(x, x)
+	if err != nil {
+		return 0, err
+	}
+	return info.Count, nil
+}
+
+// Derive implements SyncBase.
+func (ssb *SetSyncBase) Derive(p p2p.Peer) Syncer {
+	ssb.mtx.Lock()
+	defer ssb.mtx.Unlock()
+	return &setSyncer{
+		SetSyncBase: ssb,
+		OrderedSet:  ssb.os.Copy(true).(OrderedSet),
```

Comment:
There's no guarantee indeed the set doesn't change during the sync. The recent sync mechanism takes in received items from peers immediately, so if every peer uses the same set it will change during the sync.
Moreover, the original set will use `Advance()` from time to time to take in items received via gossip.
The copies are `O(1)` in terms of time and memory cost due to the use of a persistent data structure FPTree (#6406).

Reply:
I am tempting to say that it is more efficient to use copy-on-write in this case, rather than copy-on-read. that way, you just copy once for every change in the set, and all the old peers can still hang on to the outdated copy (the gc will take care of evicting that once they're all gone).

Reply:
IoW: `DBSet.Copy()` is shallow copy operation that costs next to nothing. There's a copy-on-write mechanism implemented in the underlying FPTree structure.